HINT: add hints for chained method calls

Kobzol opened the pull request:
This PR adds support for chained method hints.

![image](https://user-images.githubusercontent.com/4539057/80112702-b0b2ba80-8581-11ea-9942-d266e1be3cc8.png)

Questions:
1) ~I copied the new (201) inlay tests from `RsInlayTypeHintsProviderTestBase`, should I also add support for 193?~ Turns out that CI won't pass without it, so I added it. Also I'm not sure that I understand the syntax for writing hint tests :-)
2) The preview text in settings does not seem to be affected by the hints. Maybe it misses the std types, can they be added somehow or do I have to prepare a standalone preview without std?
3) Should I add some settings?
4) I'm a bit confused about the difference between `RsPlainHint` and the various `InlayHintsProvider`s. Should I add support for this hint to `RsPlainHint`?

Fixes: https://github.com/intellij-rust/intellij-rust/issues/5190

sirgl commented:
2. Any cross-file functionality not working in preview. In such cases you can just return null.

From my side I can say that chain hints are not very useful when they are the same across all the chain. You may check https://github.com/JetBrains/intellij-community/blob/master/java/java-impl/src/com/intellij/codeInsight/hints/MethodChainsInlayProvider.kt on how it is done for Java.

Undin commented on src/193/test/kotlin/org/rust/ide/hints/RsChainMethodHintsProviderTest.kt:
```diff
@@ -0,0 +1,17 @@
+
+class RsChainMethodHintsProviderTest : RsChainMethodHintsProviderTestBase() {
+    override fun checkInlays() {
+        myFixture.testInlays(
+            { (it.renderer as PresentationRenderer).presentation.toString() },
+            { it.renderer is PresentationRenderer }
+        )
+    }
```
Let's make some refactoring in hint tests to avoid code duplication
For example, you can create the following hierarchy:
`RsInlayTypeHintsTestPlatformBase` with single protected method `checkInlays`. Of course, you need two versions of it because of slightly different API in 2019.3 and 2020.
`RsInlayTypeHintsTestBase : RsInlayTypeHintsTestPlatformBase` with protected method `checkByText`. Right now we have two identical versions of it in `RsInlayTypeHintsProviderTestBase` and `RsChainMethodHintsProviderTestBase `
`RsInlayTypeHintsProviderTest : RsInlayTypeHintsTestBase` with all tests related to type hints (right now the corresponding class named `RsInlayTypeHintsTestBase`)
`RsChainMethodHintsProviderTest : RsInlayTypeHintsTestBase` with tests related to method chain hints
If you have ideas how to make it simpler and suggestions about better names - welcome

Kobzol commented:
For me the preview does not seem to be working at all, even if I use custom types, and even the preview for the normal type hints stopped working for me for some reason once I added the chain hints. Also (probably) my commits somehow broke a test in the normal type provider, although I did not modify that at all..

I rewrote the hinting logic to find the last element in a call chain, then traverse the chain and add hints if a type change happens. Currently the last method in the chain does not have hints (as they could be duplicated with hints in a `let` declaration). Maybe an option to enable the hint for the last method in a chain could be added into the settings?

Multiple calls on a single line look like this:
![image](https://user-images.githubusercontent.com/4539057/80194419-a395da00-861a-11ea-9433-44b06f383b06.png)
I like it better than the potential alternative, which would be something like
```rust
foo.map_err().unwrap() :Result<Option<i32>, u32>> :Option<i32>
```

I refactored the 193/201 specific code.

Kobzol commented:
For some reason, the preview cannot handle code that is otherwise type inferred and hinted correctly in a normal file.
```rust
struct A;
impl A {
    fn clone(self) -> A {
        self
    }
    fn change(self) -> B {
        B
    }
}

struct B;
impl B {
    fn clone(self) -> B {
        self
    }
}

fn foo(value: A) -> B {
    value.clone() // :A
        .clone()
        .change() // :B
        .clone()
}
```
This code should be completely standalone. In a normal file, it shows the hints, in preview it only shows `:?`. Type inference seems to be broken in the preview for some reason.

Undin commented on src/main/kotlin/org/rust/ide/hints/RsChainMethodTypeHintsProvider.kt:
```diff
@@ -0,0 +1,115 @@
+
+            override fun collect(element: PsiElement, editor: Editor, sink: InlayHintsSink): Boolean {
+                if (DumbService.getInstance(file.project).isDumb) return true
+                if (element !is RsElement) return true
```
```suggestion
                if (element !is RsMethodCall) return true
```
You need only method calls so you can check the necessary condition directly

Undin commented on src/main/kotlin/org/rust/ide/hints/RsChainMethodTypeHintsProvider.kt:
```diff
@@ -0,0 +1,115 @@
+    class Settings
+
+    companion object {
+        private val KEY: SettingsKey<Settings> = SettingsKey("rust.chain-method.hints")
```
`rust` prefix looks redundant - `SettingsKey` add it itself. See SettingsKey#getFullId`

Kobzol replied:
`RsInlayTypeHintsProvider` uses `"rust.type.hints"`, that's why I have included it. Should I remove the prefix?

Undin replied:
Yep, it was a mistake but if you remove prefix in `RsInlayTypeHintsProvider ` it will affect all users who changed the corresponding settings so I think we shouldn't do it

Undin commented on src/main/kotlin/org/rust/ide/hints/RsChainMethodTypeHintsProvider.kt:
```diff
@@ -0,0 +1,115 @@
+
+    override val name: String get() = "Chain method hints"
+
+    override val previewText: String
```
Preview doesn't work because our type inference doesn't work properly for dummy files.
I think we can omit preview for now, make some changes in type inference and after that enable it

Undin commented on src/main/kotlin/org/rust/ide/hints/RsChainMethodTypeHintsProvider.kt:
```diff
@@ -0,0 +1,115 @@
+        }, left = 1
+    )
+
+    class Settings
```
Let's add settings similar to `Minimal unique type count to show hints` Java setting

Kobzol replied:
I'm not sure if we should copy the specific behaviour of the Java setting, IMO it behaves in a weird way, at least for Rust. For example with minimal unique types = 1 (which is the lowest value that you can set it to):
```java
a
.b()
.c();
```
This doesn't show any hints. However if `b()` returned some complex type, I think that it would still be nice to show it in Rust. In Java it's mostly for builders where the types are usually obvious, while in Rust I feel like the types are often pretty complicated and you can't even name them by yourself, so you do things like assigning them into variables of a different type just to see what the real type is from the compiler error.

Also when a type repeats it suddently stops showing types:
```java
a
 .a()
 .b()
 .b()
 .c();
```

We already ignore repeated types, so some of the noise will be reduced. Maybe we could include a setting to only hint a specific type once in a specific chain? Like this:
```rust
// include_once=false
a
.a() // hint
.b() // hint
.a() // hint
.b() // hint

// include_once=true
a
.a() // hint
.b() // hint
.a()
.b()
```

Kobzol replied:
Maybe we can also ignore obvious stuff like `.clone()`? But that would be pretty brittle I think.

Undin replied:
But I understood that we should have another option. Right now our implementation doesn't show type of method call if it's the same as previous type.
At the same time, we want to hide type hints for unknown types. Since our name resolution and type inference are not ideal (unfortunately), sometimes we can get a situation when it's almost impossible from the user side to distinguish "same as previous type" and "unknown type" cases.

So let's add an option to show consecutive same method types and enable it by default - it will help us to collect a lot of feedback about our type inference. After we implemented type hints, we started to get much more reports about problems in our name resolution/type inference

Undin commented on src/main/kotlin/org/rust/ide/hints/RsChainMethodTypeHintsProvider.kt:
```diff
@@ -0,0 +1,115 @@
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.hints
```
```suggestion
@file:Suppress("UnstableApiUsage")

package org.rust.ide.hints
```

Warnings about unstable API are quite annoying

Undin commented on src/main/kotlin/org/rust/ide/hints/RsChainMethodTypeHintsProvider.kt:
```diff
@@ -0,0 +1,109 @@
+    }
+}
+
+private val RsMethodCall.isLastInChain
```
Could you add explicit returning type for all new methods and properties?

vlad20012 commented on src/main/kotlin/org/rust/ide/hints/RsChainMethodTypeHintsProvider.kt:
```diff
@@ -0,0 +1,109 @@
+                val chain = collectChain(element)
+                var lastType: Ty? = null
+                for (call in chain.dropLast(1)) {
+                    if (call.type != lastType && call.isLastOnLine) {
```
I think `!=` is ok in this case

Kobzol commented:
I added an option to show consecutive same types. Is there some way to control the hint's settings in unit tests? Or should I add some mock that will be used in unit tests to test the setting both on and off?

Comments and whitespace at the end of the line is now handled too. Is [this](https://github.com/intellij-rust/intellij-rust/pull/5276/commits/308cf387464669dcf43fe50bcf2c2ca3868cf64e#diff-a0f94161207caf4204b7980e6a4a1d90R119) safe? I assume that there are no cycles between the siblings.

Undin commented:
> Is there some way to control the hint's settings in unit tests?

I suppose you may use `com.intellij.codeInsight.hints.InlayHintsSettings#storeSettings` to set desired settings. But don't forget to restore default settings on the end of test not to affect other tests

Kobzol commented:
Thanks, I used this API to test both variants of the settings. I also rebased onto master to fix merge conflicts.

Undin commented on src/main/kotlin/org/rust/ide/hints/RsChainMethodTypeHintsProvider.kt:
```diff
@@ -0,0 +1,140 @@
+    override val previewText: String? = null
+
+    override fun createConfigurable(settings: Settings): ImmediateConfigurable = object : ImmediateConfigurable {
+        val showSameConsecutiveTypes = "Show same consecutive types"
```
Let's inline this label

Undin commented on src/main/kotlin/org/rust/ide/hints/RsChainMethodTypeHintsProvider.kt:
```diff
@@ -0,0 +1,140 @@
+            val typeHintsFactory = RsTypeHintsPresentationFactory(factory, true)
+
+            override fun collect(element: PsiElement, editor: Editor, sink: InlayHintsSink): Boolean {
+                if (DumbService.getInstance(file.project).isDumb) return true
```
```suggestion
                if (DumbService.isDumb(element.project)) return true
```

Undin commented on src/main/kotlin/org/rust/ide/hints/RsChainMethodTypeHintsProvider.kt:
```diff
@@ -0,0 +1,140 @@
+                val chain = collectChain(element)
+                var lastType: Ty? = null
+                for (call in chain.dropLast(1)) {
+                    if (call.type !is TyUnknown && call.isLastOnLine) {
```
```suggestion
                    if (call.type != TyUnknown && call.isLastOnLine) {
```

Undin commented on src/test/kotlin/org/rust/ide/hints/RsChainMethodTypeHintsProviderTest.kt:
```diff
@@ -0,0 +1,129 @@
+            service.storeSettings(key, RsLanguage, settings)
+            checkByText(code)
+        }
+        finally {
```
nit: [formatting](https://kotlinlang.org/docs/reference/coding-conventions.html#formatting-control-flow-statements)

Kobzol commented:
bors r+
